python/descriptor.c

```diff
@@ -38,113 +42,919 @@
+  size_t size;
+  char* pb = upb_encode(proto, layout, arena, &size);
+  PyObject* str = PyBytes_FromStringAndSize(pb, size);
+  // Disabled until python/message.c is reviewed and checked in.
+  // PyObject *ret = PyUpb_CMessage_MergeFromString(py_proto, str);
+  upb_arena_free(arena);
+  Py_DECREF(str);
+  return PyErr_Format(PyExc_NotImplementedError, "Not yet implemented");
+}
+
+static void PyUpb_DescriptorBase_Dealloc(PyUpb_DescriptorBase* base) {
   PyUpb_ObjCache_Delete(base->def);
   Py_DECREF(base->pool);
-  PyObject_Del(self);
+  Py_XDECREF(base->options);
+  PyUpb_Dealloc(base);
+}
+
+#define DESCRIPTOR_BASE_SLOTS                           \
+  {Py_tp_new, (void*)&PyUpb_Forbidden_New}, {           \
+    Py_tp_dealloc, (void*)&PyUpb_DescriptorBase_Dealloc \
+  }
+
+// -----------------------------------------------------------------------------
+// Descriptor
+// -----------------------------------------------------------------------------
+
+PyObject* PyUpb_Descriptor_Get(const upb_msgdef* m) {
+  const upb_filedef* file = upb_msgdef_file(m);
+  return PyUpb_DescriptorBase_Get(kPyUpb_Descriptor, m, file);
+}
+
+PyObject* PyUpb_Descriptor_GetClass(const upb_msgdef* m) {
+  PyObject* ret = PyUpb_ObjCache_Get(upb_msgdef_layout(m));
+  assert(ret);
+  return ret;
+}
+
+// The LookupNested*() functions provide name lookup for entities nested inside
+// a message.  This uses the symtab's table, which requires that the symtab is
+// not being mutated concurrently.  We can guarantee this for Python-owned
+// symtabs, but upb cannot guarantee it in general for an arbitrary
+// `const upb_msgdef*`.
+
+static const void* PyUpb_Descriptor_LookupNestedMessage(const upb_msgdef* m,
+                                                        const char* name) {
+  const upb_filedef* filedef = upb_msgdef_file(m);
+  const upb_symtab* symtab = upb_filedef_symtab(filedef);
+  PyObject* qname = PyUnicode_FromFormat("%s.%s", upb_msgdef_fullname(m), name);
+  const upb_msgdef* ret =
+      upb_symtab_lookupmsg(symtab, PyUnicode_AsUTF8AndSize(qname, NULL));
+  Py_DECREF(qname);
+  return ret;
+}
+
+static const void* PyUpb_Descriptor_LookupNestedEnum(const upb_msgdef* m,
+                                                     const char* name) {
+  const upb_filedef* filedef = upb_msgdef_file(m);
+  const upb_symtab* symtab = upb_filedef_symtab(filedef);
+  PyObject* qname = PyUnicode_FromFormat("%s.%s", upb_msgdef_fullname(m), name);
+  const upb_enumdef* ret =
+      upb_symtab_lookupenum(symtab, PyUnicode_AsUTF8AndSize(qname, NULL));
+  Py_DECREF(qname);
+  return ret;
+}
+
+static const void* PyUpb_Descriptor_LookupNestedExtension(const upb_msgdef* m,
+                                                          const char* name) {
+  const upb_filedef* filedef = upb_msgdef_file(m);
+  const upb_symtab* symtab = upb_filedef_symtab(filedef);
+  PyObject* qname = PyUnicode_FromFormat("%s.%s", upb_msgdef_fullname(m), name);
```

Comment:
why do you use PyUnicode_FromFormat instead of something lower level like sprintf?

Reply:
Yes `clang-format` did this. Fixed manually.